sandbox/pc2sandbox.sh

```diff
@@ -0,0 +1,234 @@
+# For now, just use userid
+PC2_SANDBOX_CGROUP_PATH=$PC2_CGROUP_PATH/sandbox_$USER
+
+# control whether the script outputs debug/tracing info
+_DEBUG="on"   # change this to anything other than "on" to disable debug/trace output
+DEBUG_FILE=sandbox.log
+function DEBUG()
+{
+  [ "$_DEBUG" == "on" ] && $@ >> $DEBUG_FILE
+}
+
+# ------------------------------------------------------------
+
+usage()
+{
+  cat <<SAGE
+Usage: pc2_sandbox.sh memlimit timelimit command command_args
+
+memlimit, in MB
+timelimit, in seconds
+
+SAGE
+}
+
+# Function to handle getting killed by PC2's execute timer (basically, this
+# is wall-time exceeded which is execute time limit + 1 second
+HandleTerminateFromPC2()
+{
+	DEBUG echo "Received TERMINATE signal from PC2"
+	if test -n "$submissionpid"
+	then
+		DEBUG echo "Killing off submission process $submissionpid"
+		kill -9 "$submissionpid"
+	fi
+	DEBUG echo $0: Wall time exceeded - exiting with code $FAIL_WALL_TIME_LIMIT_EXCEEDED
+	exit $FAIL_WALL_TIME_LIMIT_EXCEEDED 
+}
+
+# ------------------------------------------------------------
+
+if [ "$#" -lt 1 ] ; then
+   echo $0: No command line arguments
+   exit $FAIL_NO_ARGS_EXIT_CODE
+fi 
+
+if [ "$#" -lt 3 ] ; then
+   echo $0: expected 3 or more arguments, found: $*
+   exit $FAIL_INSUFFICIENT_ARGS_EXIT_CODE
+fi 
+
+if [ "$1" = "-h" -o "$1" = "--help" ] ; then
+   usage
+   exit $FAIL_EXIT_CODE
+fi 
+
+MEMLIMIT=$1
+TIMELIMIT=$2
+COMMAND=$3
+shift
+shift
+shift
+
+#### Debugging - just set expected first 3 args to: 16MB 5seconds
+###MEMLIMIT=8
+###TIMELIMIT=2
+###COMMAND=$1
+###shift
+
+# the rest is the command args
+
+# make sure we have CGroups V2 properly installed on this system, including a PC2 structure
+
+DEBUG echo checking PC2 CGroup V2 installation...
+if [ ! -d "$PC2_CGROUP_PATH" ]; then
+   echo $0: expected pc2sandbox CGroups v2 installation in $PC2_CGROUP_PATH 
+   exit $FAIL_INVALID_CGROUP_INSTALLATION
+fi
+
+if [ ! -f "$CGROUP_PATH/cgroup.controllers" ]; then
+   echo $0: missing file cgroup.controllers in $CGROUP_PATH
+   exit $FAIL_MISSING_CGROUP_CONTROLLERS_FILE
+fi
+
+if [ ! -f "$CGROUP_PATH/cgroup.subtree_control" ]; then
+   echo $0: missing file cgroup.subtree_control in $CGROUP_PATH
+   exit $FAIL_MISSING_CGROUP_SUBTREE_CONTROL_FILE
+fi
+
+# make sure the cpu and memory controllers are enabled
+if ! grep -q -F "cpu" "$CGROUP_PATH/cgroup.subtree_control"; then
+   echo $0: cgroup.subtree_control in $CGROUP_PATH does not enable cpu controller
+   exit $FAIL_CPU_CONTROLLER_NOT_ENABLED
+fi
+
+if ! grep -q -F "memory" "$CGROUP_PATH/cgroup.subtree_control"; then
+   echo $0: cgroup.subtree_control in $CGROUP_PATH does not enable memory controller
+   exit $FAIL_MEMORY_CONTROLLER_NOT_ENABLED
+fi
+
+
+# we seem to have a valid CGroup installation
+DEBUG echo ...done.
+
+if test -d $PC2_SANDBOX_CGROUP_PATH
+then
+	DEBUG echo Removing existing sandbox to start clean
+	rmdir $PC2_SANDBOX_CGROUP_PATH
```

Comment:
Is it possible that the $PC2_SANDBOX_CGROUP_PATH directory will have files and/or subdirs in it?  If so, won't this `rmdir` fail?

Reply:
In fact, it **will** have "virtual" files in it.  There should not be "real" files in it.  If you '`rmdir`' a cgroup folder, it kills the whole cgroup under it, unless it is inuse, which it shouldn't be since the guy running it is the one removing it.

There is a comment at the top:
```
# TODO: should have unique runbox in case multiple instances are executing in separate judges
# For now, just use userid

```If two people log in using the same userid, and attempt to judge on the same machine, it will cause problems.
It's  TODO because I'm not sure how to address that.